Fixed ‘vs’ command by:

palcalde opened the pull request:
Fixed ‘vs’ command by (Issue https://github.com/facebook/chisel/issues/88#issuecomment-98440239):

- Stop using FBInputHandler. This class uses SBInputReader which no
longer exist in the lldb version that comes bundled with XCode (that’s
why the command was broken), even tho it does exist in the official
lldb documentation
(http://lldb.llvm.org/python_reference/lldb.SBInputReader-class.html).

- Now the script uses the standard stdin library to read from terminal.

- Removed the ‘Flickering’ of the selected view, instead we now use
mask which is a lot more useful visually.

NOTE: As mentioned this fix doesn't use the class FBInputHandler at all to avoid SBInputReader, but if we have that class back in next versions of XCode we can always revert this fix-commit.

kastiglione commented on commands/FBFlickerCommands.py:
```diff
@@ -116,13 +112,16 @@ def inputCallback(self, input):
     viewHelpers.setViewHidden(oldView, False)
 
-  def setCurrentView(self, view):
+  def setCurrentView(self, view, oldView):
```
Default `oldView` to `None`?

kastiglione commented on commands/FBFlickerCommands.py:
```diff
@@ -62,20 +63,14 @@ def run(self, arguments, options):
-    viewHelpers.setViewHidden(self.currentView, False)
-    time.sleep(0.3)
+    self.keepRunning = True
+    lldb.debugger.SetAsync (True)
```
Can you comment on why this is necessary? And, remove the space after `SetAsync`?

kastiglione commented:
@palcalde Thanks!

I've left a couple comments.

Also, can we make the mask a flag? (You can see examples of how to do flags in other chisel commands.) I'd prefer not to arbitrarily change the command from flickering to masking. Ideally, this could also use a border instead of flicker or mask, but I'm not asking you to go that far :stuck_out_tongue_winking_eye: .

palcalde commented:
@kastiglione thanks for reviewing. I will add a comment about the async mode. The reason is because if you are on lldb console and type vs, the script blocks the thread waiting for user input. That is expected, however, if you type on continue in the debugger XCode enables the simulator again but lldb becomes absolutely blocked. You can't stop the execution or anything like that. The only thing you can do is quit simulator to kill the lldb process. 

I found that if you run in async mode, lldb runs in a diff thread so it doesn't become blocked when you continue debugger in vs mode. You can stop again and keep using it normally. 

About the border instead of masking, I will investigate and try to see if I can put it in the open PR.

kastiglione commented:
> About the border instead of masking, I will investigate and try to see if I can put it in the open PR.

To be clear, it's not important to add border support. I'm only concerned about keeping flicker support. I'd rather not remove flicker support, but if we want to augment it with mask support, :+1:.

For async mode, are there consequences to leaving it in async mode? Or should it be returned to off after `vs` completes?

palcalde commented:
@kastiglione The only reason I moved flickering to masking is because the new implementation doesn't use SBInputReader, it uses stdin which blocks the thread waiting for user input. SBInputReader was a separated buffer so the previous implementation did a while hiding/showing the view every 0.1 sec. 

Since adding an async implementation with stdin in python is not trivial (it would require a Queue a Thread and a couple of methods) I just thought it wasn't worth it and just went for a simple mask behavior. That is the only reason why I change it :)

I added one more commit fixing the issues you mentioned.

kastiglione commented on commands/FBFlickerCommands.py:
```diff
@@ -62,20 +63,17 @@ def run(self, arguments, options):
-
-    self.handler = inputHelpers.FBInputHandler(lldb.debugger, self.inputCallback)
-    self.handler.start()
+    self.setCurrentView(startView, None)
```
Now that `oldView` defaults to `None`, the `None` here can be removed.

kastiglione commented:
Thanks for the explanation. Let's do it. A masking `vs` is better than a broken `vs`.

I have one more comment, otherwise looks good and thank you so much for the pull request!

palcalde commented:
Sure, no worries it was fun dealing with lldb and Python a little bit! :+1: 
I will update the pull request with your comment.

palcalde commented:
Pull Request updated.

kastiglione commented on commands/FBFlickerCommands.py:
```diff
@@ -116,13 +115,14 @@ def inputCallback(self, input):
-      print '\nI really have no idea what you meant by \'' + input + '\'... =\\\n'
-
-    viewHelpers.setViewHidden(oldView, False)
+      print '\nChisel - VS Mode: I really have no idea what you meant by \'' + input + '\'... =\\\n'
```
Just noticed this. We don't really do this anywhere else with any output. Do you think it's necessary to mention Chisel and VS by name?

palcalde commented:
The only reason why I put that in the message is because is the only command in Chisel (that I know) that stops lldb waiting for user input. 

I thought it could be useful to remind people they are in vs mode. 

I can reverse it if you feel more comfortable. Just let me know!

palcalde commented:
@kastiglione reverted message so is consistent with the rest of commands.

kastiglione commented:
:fireworks:
